src/_guides/language/concurrency/index.md

```diff
@@ -0,0 +1,456 @@
+
+```dart
+Future _readAndParseJson(SendPort p) async {
+  final fileData = await File(filename).readAsString();
+  final jsonData = jsonDecode(fileData);
+  Isolate.exit(p, jsonData);
+}
+```
+
+The relevant statement is the last one, which exits the isolate,
+sending `jsonData` to the passed-in `SendPort`.
+Message passing between isolates normally involves data copying,
+and thus can be slow and increases with the size of the message —
+O(n) in big O notation.
+However, when you send the data using `Isolate.exit()`,
+then the memory that holds the message in the exiting isolate isn’t copied,
+but instead is transferred to the receiving isolate.
+That transfer is quick and completes in constant time — O(1).
+
+{{site.alert.version-note}}
+  `Isolate.exit()` was added in 2.15.
+  Previous releases support only explicit message passing,
+  using `Isolate.send()` as shown in the next section's example.
+{{site.alert.end}}
+
+The following figure summarizes the communication between
+the main isolate and the second isolate:
+
+![PENDING: alt text here](/guides/language/concurrency/images/isolate-api.png)
+<br>
+***[PENDING: Inside the figure, Background worker -> Worker isolate]***
+
+
+### Sending multiple messages between isolates
+
+If you need more communication between isolates,
+then you need to use the [`send()` method][] of `SendPort`.
+One common pattern, which the following figure shows,
+is for the main isolate to send a request message to the worker isolate,
+which then sends one or more reply messages.
+
+![PENDING: alt text here](/guides/language/concurrency/images/isolate-custom-bg-worker.png)
+<br>
+***[PENDING: The top arrow should be labeled "Create isolate: Isolate.spawn()"]***
+<br>
+***[PENDING: Inside the figure, Custom background worker -> Worker isolate]***
+
+
+For examples of sending multiple messages, see the following:
+
+* ***[PENDING: link to example 4 in Isolate snippets][]***,
+  which shows how to send a message from
+  the main isolate to the spawned isolate.
+  It’s otherwise similar to the preceding example.
+* ***[PENDING: link to example 5 in Isolate snippets][]***,
+  which shows how to spawn a long-running isolate that
+  receives and sends multiple times.
+
+{{site.alert.tip}}
+  The `send()` method can be slow.
+  When possible,
+  avoid using `send()` to send large amounts of data.
+  Instead, send the data when the worker isolate exits,
+  as described in the previous section.
+{{site.alert.end}}
+
+## Performance and isolate groups
+
+When an isolate calls [`Isolate.spawn()`][],
+the two isolates have the same executable code
+and are in the same _isolate group_.
+Isolate groups enable performance optimizations such as sharing code;
+a new isolate immediately runs the code owned by the isolate group.
+Also, `Isolate.exit()` works only when the isolates
+are in the same isolate group.
+
+In some special cases —
+for example, when implementing tools like `dart pub run` —
```

Comment:
I'm not sure the reference to "implementing tools like dart pub run" is all that helpful here. Could we just say something like "for example, when you need to spawn a new isolate from new source code read from a Uri"?

Reply:
That explanation is already in the next sentence, so I think I'll just delete the example.